### PR TITLE
Name resolution refactoring

### PR DESCRIPTION
**Pull request by flodiebold**
This is still very WIP, but it's becoming quite big and I want to make sure this isn't going in a completely bad direction :sweat_smile:. I'm not really happy with how the path resolution looks, and I'm not sure `PerNs<Resolution>` is the best return type -- there are 'this cannot happen in the (types/values) namespace' cases everywhere. I also want to unify the `resolver` and `nameres` namespaces once I'm done switching everything to `Resolver`. Also, `Resolver` only has a lifetime because it needs to have a reference to the `ItemMap` during import resolution :confused: 

The differences in the completion snapshots are almost completely just ordering (except it completes `Self` as well now), so I changed it to sort the completions before snapshotting.

### CODE REVIEW

**Comment by matklad** [comment]
>The differences in the completion snapshots are almost completely just ordering (except it completes Self as well now), so I changed it to sort the completions before snapshotting.

Let’s file it in a separate pr perhaps to minimize the diff?

**Comment by matklad** [comment]
We might also look at:

https://github.com/brendanzab/moniker

I honestly don’t understand what this library is doing exactly, but it seems cool and I am all for using libraries :-)

**Comment by flodiebold** [comment]
> Let’s file it in a separate pr perhaps to minimize the diff?

Good point, I already made it a separate commit so I just need to reorder them a bit...

**Comment by flodiebold** [comment]
> We might also look at:
> 
> https://github.com/brendanzab/moniker
> 
> I honestly don’t understand what this library is doing exactly, but it seems cool and I am all for using libraries :-)

Yeah, it seems complicated :wink:

File: crates/ra_hir/src/resolve.rs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    item_map: Arc<ItemMap>,
+    module: Module,
+}
+
+/// For usage during import resolution.
+#[derive(Debug, Clone)]
+pub(crate) struct ModuleItemMapRef<'a> {
+    item_map: &'a ItemMap,
+    module: Module,
+}
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]: These two could be ModuleItemMap<T: AsRef<ItemMap>>?

File: crates/ra_hir/src/resolve.rs [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+    pub fn resolve_path(&self, db: &impl HirDatabase, path: &Path) -> PerNs<Resolution> {
+        self.resolve_path_fp(db, path).0
+    }
+
+    pub(crate) fn resolve_path_fp(
+        &self,
+        db: &impl HirDatabase,
+        path: &Path,
+    ) -> (PerNs<Resolution>, ReachedFixedPoint) {
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]: Conversely, talking about FixedPoint in general resolutions seems confusing

File: crates/ra_hir/src/resolve.rs [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    },
+    /// A local binding (only value namespace)
+    LocalBinding {
+        pat: PatId,
+    },
+    /// A generic parameter
+    GenericParam {
+        idx: u32,
+    },
+    SelfType(ImplBlock),
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]: Self type feels like a kind of generic param

File: crates/ra_hir/src/resolve.rs [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    res: (PerNs<ModuleDef>, ReachedFixedPoint),
+) -> (PerNs<Resolution>, ReachedFixedPoint) {
+    (res.0.map(|def| Resolution::Def { def }), res.1)
+}
+
+fn resolve_path_segments(
+    db: &impl HirDatabase,
+    // during import resolution for a certain, we can't use the query to get the
+    // item map for that crate since it's just being constructed. So in that
+    // case, we need to pass the item map here.
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]: Yeah, this is this awkward bit of “an abstraction interface has two sides: “public”, for consumers, and “priviate”, for implemenetrs. It sucks when one is round and he other is squared”

**Comment by matklad** [comment]
Terminology note: I use descriptors(kotlin)/symbols(dotty and roslyn)/items(rustc) and defs(rustc) somewhat interchangeably. I suggest using the Def as a general term, as the shortest and least ambiguous. :)

**Comment by flodiebold** [comment]
You're right, trying to unify the path resolution like that is the wrong approach. I think wrapping and extending the path resolution of the `ItemMap` will probably work better.

File: crates/ra_hir/src/resolve.rs (hunk 3 above)
Comment (flodiebold) [reply crates/ra_hir/src/resolve.rs]: Yeah, I think that's how rustc handles it as well. Currently we don't handle impl block generics at all though, I think when we implement that, we can do it that way.

**Comment by matklad** [comment]
One important though (which is present in the code, but not documented):

We should only store *scopes* themselves, and avoid storing def-use links: there are many def-use links than scopes, and just resolving use every-time using precomputed scopes should not be that costly.

File: crates/ra_hir/src/resolve.rs [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    name::{Name, KnownName},
+    nameres::{PerNs, ItemMap, ReachedFixedPoint},
+    generics::GenericParams,
+    expr::{scope::{ExprScopes, ScopeId}, PatId},
+    impl_block::ImplBlock,
+    path::{Path, PathKind, PathSegment},
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct Resolver<'a> {
```
[the remark this answers is not in the document]
Comment (flodiebold) [reply crates/ra_hir/src/resolve.rs]: It's a cool name, but I think I like `Resolver` more because it focuses on what the user does with it :)

**Comment by matklad** [comment]
Love this interface!

bors r+

See also 
https://github.com/rust-analyzer/rust-analyzer/issues/723 for some required changes in the implementation to support macros properly.

**Pull request merged**